moped-editor/src/views/projects/projectView/KnackSync.js

```diff
@@ -0,0 +1,217 @@
+import React from "react";
+import { MenuItem, Icon, ListItemIcon, ListItemText } from "@material-ui/core";
+
+import { useMutation } from "@apollo/client";
+
+import { UPDATE_PROJECT_KNACK_ID } from "../../../queries/project";
+
+/**
+ * Entry in project menu drop down to trigger sync to Data Tracker
+ * @return {JSX.Element}
+ * @constructor
+ */
+const KnackSync = React.forwardRef(
+  (
+    {
+      project,
+      closeHandler,
+      snackbarHandler,
+      sceneNumber = 514,
+      viewNumber = 3047,
+      knackApplicationId = "6167314778435d001ea3e7cb",
+    },
+    ref
+  ) => {
+    /**
+     * Function to build the correct Knack URL to interact with based on properties and if there will be an
+     * update or an initial sync.
+     * @returns string
+     */
+    const buildUrl = () => {
+      let url =
+        "https://api.knack.com/v1/pages/scene_" +
+        sceneNumber +
+        "/views/view_" +
+        viewNumber +
+        "/records";
+      if (project.moped_project[0].knack_project_id) {
+        // existing record
+        url = url + "/" + project.moped_project[0].knack_project_id;
+      }
+      return url;
+    };
+
+    /**
+     * Function to determine the HTTP method to use base on if there will be an update or initial post to Knack
+     * @returns string
+     */
+    const getHttpMethod = () => {
+      let method = "POST";
+      if (project.moped_project[0].knack_project_id) {
+        method = "PUT";
+      }
+      return method;
+    };
+
+    /**
+     * Function to build up the headers which need to be sent as part of the a Knack API call
+     * @returns object
+     */
+    const buildHeaders = () => {
+      let headers = {
+        "Content-Type": "application/json",
+        "X-Knack-Application-Id": knackApplicationId,
+        "X-Knack-REST-API-Key": "knack",
+      };
+      return headers;
+    };
```

Comment:
agree and am not sure where to build in a protection against WSOD. what do you suggest @sergiogcx ?

Reply:
@johnclary I was thinking of something like this maybe:

```
const getHttpMethod = (project.moped_project?.[0]?.knack_project_id ?? false) ? "PUT" : "POST;
```

Reply:
* Totally agree, the header object needed is the same for the initial sync HTTP call and for all following syncs. I've defined it as an object and not a function - great catch.

* @sergiogcx's suggested code here is really sharp and exactly what we need to get the right HTTP method. I don't think we can get away with computing the value once though and hanging on to it as a const because it's going to be a POST for the initial sync and then PUTs for all following calls. Having said that, there may be possible to rely on react to tear down the component when the menu closes and recompute the correct HTTP method when the component is instantiated when the menu is opened again.

I'll do some testing to see if that's how things work - please let me know if there's something I'm missing.